webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java

```diff
@@ -0,0 +1,252 @@
+            translationsBySourceText = leveragedTranslationsBySourceText;
+
+            // Only MT strings for which we didn't leverage translationsBySourceText for all the expected target languages.
+            // Note: this would benefit from further optimization at some point to take into account partially leveraged
+            // text units - though that would require additional MT API calls for certain MT engine implementations.
+            if (!leveragedTranslationsBySourceText.isEmpty()) {
+                untranslatedStrings = textSources.stream()
+                        .filter(sourceText -> !leveragedTranslationsBySourceText.containsKey(sourceText)
+                                || leveragedTranslationsBySourceText.get(sourceText).size() != targetBcp47Tags.size())
+                        .collect(Collectors.toList());
+            }
+        }
+
+        ImmutableMap<String, ImmutableList<TranslationDTO>> machineTranslationsBySourceText = getMachineTranslationBySourceText(
+                getSourceBcp47TagOrDefault(sourceBcp47Tag),
+                targetBcp47Tags,
+                untranslatedStrings, skipFunctionalProtection);
+
+        translationsBySourceText = translationMerger.getMergedTranslationsBySourceText(translationsBySourceText, machineTranslationsBySourceText);
+        return getSortedTranslationsResponse(textSources, targetBcp47Tags, translationsBySourceText);
+    }
+
+    public TranslationDTO getSingleTranslation(
+            String textSource,
+            String sourceBcp47Tag,
+            String targetBcp47Tags,
+            boolean skipFunctionalProtection,
+            boolean skipLeveraging,
+            ArrayList<Long> repositoryIds,
+            ArrayList<String> repositoryNames
+    ) {
+        return getTranslations(
+                Collections.singletonList(textSource),
+                sourceBcp47Tag,
+                Collections.singletonList(targetBcp47Tags),
+                skipFunctionalProtection,
+                skipLeveraging,
+                repositoryIds,
+                repositoryNames)
+                .getTextUnitTranslations().get(0)
+                .getTranslations().get(0);
+    }
+
+    @Timed("MachineTranslationService.getMachineTranslatedResponse")
+    ImmutableMap<String, ImmutableList<TranslationDTO>> getMachineTranslationBySourceText(String sourceBcp47Tag, List<String> targetBcp47Tags, List<String> textSources, Boolean skipFunctionalProtection) {
+        if (textSources.isEmpty()) {
+            return ImmutableMap.of();
+        }
+
+        // TODO(garion): Implement functional protection / placeholder processing
+        if (skipFunctionalProtection != null && skipFunctionalProtection) {
+            logger.debug("function / placeholder protection");
+        }
+
+        return machineTranslationEngine.getTranslationsBySourceText(
+                textSources,
+                sourceBcp47Tag,
+                targetBcp47Tags,
+                "",
```

Comment:
empty strings? should that be null, Optional or some enum?

Reply:
That should be null for now. Once a source placeholder escaping mechanism is introduced, that will be replaced with a relevant enum.